### ISSUE TITLE
`UP035` false positive on `typing_extensions.dataclass_transform` on python 3.11

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#5112]
```py
from typing_extensions import dataclass_transform
# UP035 Import from `typing` instead: `dataclass_transform`
```

`typing_extensions.dataclass_transform` has the `frozen_default` parameter which won't be in `typing` until python 3.12.

from `typing_extensions`:
```py
if sys.version_info >= (3, 12):
    # dataclass_transform exists in 3.11 but lacks the frozen_default parameter
    dataclass_transform = typing.dataclass_transform
else:
    def dataclass_transform(
        *,
        eq_default: bool = True,
        order_default: bool = False,
        kw_only_default: bool = False,
        frozen_default: bool = False,
        field_specifiers: typing.Tuple[
            typing.Union[typing.Type[typing.Any], typing.Callable[..., typing.Any]],
            ...
        ] = (),
        **kwargs: typing.Any,
    ) -> typing.Callable[[T], T]:
        ...
```

**Comment by charliermarsh** [comment]
Ah interesting, okay, thank you.

**Comment by AlexWaygood** [comment]
Looks like ruff will also do this:

```diff
- from typing_extensions import SupportsIndex
+ from typing import SupportsIndex

  isinstance(42, SupportsIndex)
```

But the `isinstance()` check there is 20x faster if you use the `typing_extensions` version, as `typing_extensions` backports a ton of optimisations that have been added to `typing.Protocol` in Python 3.12

**Comment by charliermarsh** [comment]
@AlexWaygood - Would you suggest not rewriting `Protocol` imports either for the same reason?

**Comment by AlexWaygood** [comment]
> @AlexWaygood - Would you suggest not rewriting `Protocol` imports either for the same reason?

Yes -- and FWIW, pyupgrade recently made that change in https://github.com/asottile/pyupgrade/commit/5059713466c113e8aa09ab5c93b2c82f357ab206 (following my PR https://github.com/asottile/reorder-python-imports/pull/341 to the reorder-python-imports project, which pyupgrade uses).

**Comment by charliermarsh** [comment]
Thanks!
